File: src/main/java/com/mercateo/jsonschema/generictype/GenericType.java
package com.mercateo.jsonschema.generictype;

import com.googlecode.gentyref.GenericTypeReflector;

import java.lang.reflect.*;

import static java.util.Objects.requireNonNull;

public interface GenericType<T> {
    Class<T> getRawType();

    String getSimpleName();

    Type getType();

    GenericType<?> getContainedType();

    boolean isInstanceOf(Class<?> clazz);

    boolean isIterable();

    Field[] getDeclaredFields();

    GenericType<? super T> getSuperType();

    static GenericType<?> of(Type type) {
        return of(type, null);
    }

    @SuppressWarnings("unchecked")
    static <T> GenericType<T> of(Type type, Class<T> rawType) {
        if (type instanceof ParameterizedType) {
            ParameterizedType parameterizedType = (ParameterizedType) type;
            return new GenericParameterizedType<>(parameterizedType, (Class<T>) parameterizedType
                .getRawType());
        } else if (type instanceof Class) {
            return new GenericClass<>((Class<T>) type);
        } else if (type instanceof GenericArrayType) {
            return new GenericArray<>((GenericArrayType) type, requireNonNull(rawType));
        }
        {
            throw new IllegalStateException("unhandled type " + type);
        }
    }

    static GenericType<?> of(Field field, Type type) {
        final Class<?> fieldClass = field.getType();
        final Type fieldType = GenericTypeReflector.getExactFieldType(field, type);
        return of(fieldType, fieldClass);
    }

    static GenericType<?> of(Method method, Type type) {
        final Class<?> returnClass = method.getReturnType();
        final Type returnType = GenericTypeReflector.getExactReturnType(method, type);
        return of(returnType, returnClass);
    }

}

File: src/main/java/com/mercateo/jsonschema/generictype/GenericType.kt
package com.mercateo.jsonschema.generictype

import com.googlecode.gentyref.GenericTypeReflector

import java.lang.reflect.*

import java.util.Objects.requireNonNull

interface GenericType<out T> {
    val rawType: Class<out T>

    val simpleName: String

    val name: String

    val type: Type

    val containedType: GenericType<*>

    fun isInstanceOf(clazz: Class<*>): Boolean

    val isIterable: Boolean

    val declaredFields: Array<Field>

    val declaredMethods: Array<Method>

    val superType: GenericType<Any>?

    companion object {

        fun of(type: Type): GenericType<Any> {
            return of<Any>(type, null)
        }

        fun <T> of(type: Class<T>): GenericType<T> {
            return of(type, null)
        }

        @SuppressWarnings("unchecked")
        fun <T> of(type: Type, rawType: Class<T>?): GenericType<T> {
            if (type is ParameterizedType) {
                return GenericParameterizedType(type, type.rawType as Class<T>)
            } else if (type is Class<*>) {
                return GenericClass(type as Class<T>)
            } else if (type is GenericArrayType) {
                return GenericArray(type, requireNonNull<Class<T>>(rawType))
            }
            run { throw IllegalStateException("unhandled type " + type) }
        }

        fun ofField(field: Field, type: Type): GenericType<Any> {
            val fieldClass = field.type
            val fieldType = GenericTypeReflector.getExactFieldType(field, type)
            return of(fieldType, fieldClass)
        }

        fun ofMethod(method: Method, type: Type): GenericType<Any> {
            val returnClass = method.returnType
            val returnType = GenericTypeReflector.getExactReturnType(method, type)
            return of(returnType, returnClass)
        }
    }

}
